:farmer: Build warning: `__builtin_memmove` out of bounds in `test_buffer.cpp` under GCC 14

### Generated by Generative AI

_No response_

### Operating System:

rhel

### ROS version or commit hash:

Rolling

### RMW implementation (if applicable):

_No response_

### RMW Configuration (if applicable):

_No response_

### Client library (if applicable):

_No response_

### 'ros2 doctor --report' output

_No response_

### Steps to reproduce issue

1. Go to
2. Build a job
3. See the warning

### Expected behavior

To build without warnings

### Actual behavior

Building with warnings

### Additional information

### Reference Build
- [nightly_linux-rhel_release#2699](https://ci.ros2.org/job/nightly_linux-rhel_release/2699)

### Log Output
```
---
--- stderr: rosidl_buffer
In file included from /usr/include/c++/14/string:51,
                 from /usr/include/c++/14/bits/locale_classes.h:40,
                 from /usr/include/c++/14/bits/ios_base.h:41,
                 from /usr/include/c++/14/iomanip:42,
                 from /usr/include/gtest/gtest.h:54,
                 from /home/jenkins-agent/workspace/nightly_linux-rhel_release/ws/src/ros2/rosidl/rosidl_buffer/test/test_buffer.cpp:15:
In static member function ‘static _Up* std::__copy_move<_IsMove, true, std::random_access_iterator_tag>::__copy_m(_Tp*, _Tp*, _Up*) [with _Tp = unsigned char; _Up = unsigned char; bool _IsMove = true]’,
    inlined from ‘_OI std::__copy_move_a2(_II, _II, _OI) [with bool _IsMove = true; _II = unsigned char*; _OI = unsigned char*]’ at /usr/include/c++/14/bits/stl_algobase.h:521:30,
    inlined from ‘_OI std::__copy_move_a1(_II, _II, _OI) [with bool _IsMove = true; _II = unsigned char*; _OI = unsigned char*]’ at /usr/include/c++/14/bits/stl_algobase.h:548:42,
    inlined from ‘_OI std::__copy_move_a(_II, _II, _OI) [with bool _IsMove = true; _II = __gnu_cxx::__normal_iterator<unsigned char*, vector<unsigned char, allocator<unsigned char> > >; _OI = __gnu_cxx::__normal_iterator<unsigned char*, vector<unsigned char, allocator<unsigned char> > >]’ at /usr/include/c++/14/bits/stl_algobase.h:555:31,
    inlined from ‘_OI std::move(_II, _II, _OI) [with _II = __gnu_cxx::__normal_iterator<unsigned char*, vector<unsigned char, allocator<unsigned char> > >; _OI = __gnu_cxx::__normal_iterator<unsigned char*, vector<unsigned char, allocator<unsigned char> > >]’ at /usr/include/c++/14/bits/stl_algobase.h:683:38,
    inlined from ‘std::vector<_Tp, _Alloc>::iterator std::vector<_Tp, _Alloc>::_M_erase(iterator) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]’ at /usr/include/c++/14/bits/vector.tcc:185:2,
    inlined from ‘std::vector<_Tp, _Alloc>::iterator std::vector<_Tp, _Alloc>::erase(const_iterator) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]’ at /usr/include/c++/14/bits/stl_vector.h:1550:24,
    inlined from ‘rosidl::Buffer<T, Allocator>::iterator rosidl::Buffer<T, Allocator>::erase(const_iterator) [with T = unsigned char; Allocator = std::allocator<unsigned char>]’ at /home/jenkins-agent/workspace/nightly_linux-rhel_release/ws/src/ros2/rosidl/rosidl_buffer/include/rosidl_buffer/buffer.hpp:452:47,
    inlined from ‘virtual void TestBufferNonCpu_modifiers_throw_Test::TestBody()’ at /home/jenkins-agent/workspace/nightly_linux-rhel_release/ws/src/ros2/rosidl/rosidl_buffer/test/test_buffer.cpp:962:3:
/usr/include/c++/14/bits/stl_algobase.h:452:30: warning: ‘void* __builtin_memmove(void*, const void*, long unsigned int)’ forming offset 2 is out of the bounds [0, 2] [�]8;;https://gcc.gnu.org/onlinedocs/gcc-14.3.0/gcc/Warning-Options.html#index-Warray-bounds�-Warray-bounds=�]8;;�]
  452 |             __builtin_memmove(__result, __first, sizeof(_Tp) * _Num);
      |             ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In static member function ‘static _Up* std::__copy_move<_IsMove, true, std::random_access_iterator_tag>::__copy_m(_Tp*, _Tp*, _Up*) [with _Tp = unsigned char; _Up = unsigned char; bool _IsMove = true]’,
    inlined from ‘_OI std::__copy_move_a2(_II, _II, _OI) [with bool _IsMove = true; _II = unsigned char*; _OI = unsigned char*]’ at /usr/include/c++/14/bits/stl_algobase.h:521:30,
    inlined from ‘_OI std::__copy_move_a1(_II, _II, _OI) [with bool _IsMove = true; _II = unsigned char*; _OI = unsigned char*]’ at /usr/include/c++/14/bits/stl_algobase.h:548:42,
    inlined from ‘_OI std::__copy_move_a(_II, _II, _OI) [with bool _IsMove = true; _II = __gnu_cxx::__normal_iterator<unsigned char*, vector<unsigned char, allocator<unsigned char> > >; _OI = __gnu_cxx::__normal_iterator<unsigned char*, vector<unsigned char, allocator<unsigned char> > >]’ at /usr/include/c++/14/bits/stl_algobase.h:555:31,
    inlined from ‘_OI std::move(_II, _II, _OI) [with _II = __gnu_cxx::__normal_iterator<unsigned char*, vector<unsigned char, allocator<unsigned char> > >; _OI = __gnu_cxx::__normal_iterator<unsigned char*, vector<unsigned char, allocator<unsigned char> > >]’ at /usr/include/c++/14/bits/stl_algobase.h:683:38,
    inlined from ‘std::vector<_Tp, _Alloc>::iterator std::vector<_Tp, _Alloc>::_M_erase(iterator, iterator) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]’ at /usr/include/c++/14/bits/vector.tcc:201:6,
    inlined from ‘std::vector<_Tp, _Alloc>::iterator std::vector<_Tp, _Alloc>::erase(const_iterator, const_iterator) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]’ at /usr/include/c++/14/bits/stl_vector.h:1581:17,
    inlined from ‘rosidl::Buffer<T, Allocator>::iterator rosidl::Buffer<T, Allocator>::erase(const_iterator, const_iterator) [with T = unsigned char; Allocator = std::allocator<unsigned char>]’ at /home/jenkins-agent/workspace/nightly_linux-rhel_release/ws/src/ros2/rosidl/rosidl_buffer/include/rosidl_buffer/buffer.hpp:458:47,
    inlined from ‘virtual void TestBufferNonCpu_modifiers_throw_Test::TestBody()’ at /home/jenkins-agent/workspace/nightly_linux-rhel_release/ws/src/ros2/rosidl/rosidl_buffer/test/test_buffer.cpp:963:3:
/usr/include/c++/14/bits/stl_algobase.h:452:30: warning: ‘void* __builtin_memmove(void*, const void*, long unsigned int)’ forming offset 2 is out of the bounds [0, 2] [�]8;;https://gcc.gnu.org/onlinedocs/gcc-14.3.0/gcc/Warning-Options.html#index-Warray-bounds�-Warray-bounds=�]8;;�]
  452 |             __builtin_memmove(__result, __first, sizeof(_Tp) * _Num);
      |             ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
---
Finished <<< rosidl_buffer [16.3s]
```

This warning had a different output before May 19th, but after this pr `https://github.com/ros2/rosidl/commit/837262f5fc7fa4e45dd7ece2b3cae35c6b276e7f` this is the new output

CC: @Crola1702

--- a/rosidl_buffer/test/test_buffer.cpp
+++ b/rosidl_buffer/test/test_buffer.cpp
@@ -954,14 +954,6 @@ TEST(TestBufferNonCpu, modifiers_throw) {
   EXPECT_THROW(buffer.push_back(1), std::runtime_error);
   EXPECT_THROW(buffer.pop_back(), std::runtime_error);
   EXPECT_THROW(buffer.emplace_back(1), std::runtime_error);
-
-  using ConstIt = Buffer<uint8_t>::const_iterator;
-  std::vector<uint8_t> aux = {0, 1};
-  ConstIt it0 = aux.cbegin();
-  ConstIt it1 = aux.cbegin() + 1;
-  EXPECT_THROW(buffer.erase(it0), std::runtime_error);
-  EXPECT_THROW(buffer.erase(it0, it1), std::runtime_error);
-  EXPECT_THROW(buffer.emplace(it0, 1), std::runtime_error);
 }
 
 TEST(TestBufferNonCpu, capacity_throws) {

